libvast/builtins/connectors/file.cpp

```diff
@@ -0,0 +1,268 @@
+          return caf::make_error(ec::filesystem_error,
+                                 fmt::format("file {} is not readable", arg));
+        }
+        path = arg;
+      } else {
+        return caf::make_error(
+          ec::invalid_argument,
+          fmt::format("unexpected argument for 'file' connector: {}", arg));
+      }
+    }
+    if (path.empty()) {
+      return caf::make_error(ec::syntax_error,
+                             fmt::format("no file specified"));
+    }
+    auto fd = STDIN_FILENO;
+    switch (file_type) {
+      default:
+        return caf::make_error(ec::filesystem_error, "unsupported file type",
+                               path);
+      case std::filesystem::file_type::socket: {
+        if (path == stdin_path) {
+          return caf::make_error(ec::filesystem_error,
+                                 "cannot use STDIN as UNIX "
+                                 "domain socket");
+        }
+        auto uds = detail::unix_domain_socket::connect(path);
+        if (!uds) {
+          return caf::make_error(ec::filesystem_error,
+                                 "failed to connect to UNIX domain socket at",
+                                 path);
+        }
+        fd = uds.recv_fd(); // Blocks!
+        break;
+      }
+      case std::filesystem::file_type::regular: {
+        if (path != stdin_path) {
+          fd = ::open(path.c_str(), std::ios_base::binary | std::ios_base::in);
+          if (fd == -1) {
+            return caf::make_error(ec::filesystem_error,
+                                   "open(2) for file {} failed {}:", path,
+                                   std::strerror(errno));
+          }
+        }
+        break;
+      }
+    }
+    return std::invoke(
+      [](auto timeout, auto fd, auto following) -> generator<chunk_ptr> {
+        auto in_buf = detail::fdinbuf(fd, max_chunk_size);
+        in_buf.read_timeout() = timeout;
+        auto current_data = std::vector<std::byte>{};
+        current_data.reserve(max_chunk_size);
+        auto eof_reached = false;
+        while (following or not eof_reached) {
+          auto current_char = in_buf.sbumpc();
+          if (current_char != detail::fdinbuf::traits_type::eof()) {
+            current_data.emplace_back(static_cast<std::byte>(current_char));
+          } else {
+            eof_reached = (not in_buf.timed_out());
+            if (current_data.empty()) {
+              if (not eof_reached) {
+                co_yield chunk::make_empty();
+                continue;
+              }
+              if (eof_reached and not following) {
+                break;
+              }
+            }
+          }
+          if (eof_reached or current_data.size() == max_chunk_size) {
+            auto chunk = chunk::make(std::exchange(current_data, {}));
+            co_yield std::move(chunk);
+            if (following or not eof_reached) {
+              current_data.reserve(max_chunk_size);
+            }
+          }
```

Comment:
What exactly should happen if a timeout occurs? If I read this implementation correctly, we emit an empty chunk if we did not collect any data yet, and otherwise we just go back and try reading again. That's probably not what we want to do, or is it?

Reply:
You're right, but that's exactly how we specified the timeout should work, as a sort of "keepalive" for the next pipeline operator. Example: Receiving pipeline input data packets from remote connections. 

Has this specification been changed?

Reply:
In a Slack discussion, we agreed to make the generator always return a chunk upon hitting a timeout. https://github.com/tenzir/vast/pull/3088/commits/4158b5d2f79ca238fd72a558029d53a5f0cd2bae